rfcs/20200106-prelu/README.md

```diff
@@ -0,0 +1,137 @@
+# Introducing PReLU support in oneDNN
+
+## Introduction
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation function.
+Request comes from Caffe users: "PReLU is used in some POR topologies, such as
+VNET, SRGAN. It takes most of time if the original implementation, is not good
+optimized. For example VNET's train perf with latest Intel Caffe is 2.4s, PReLU
+(has optimized by openmp) spends 429ms (not including reorder time), 17% of total
+time. If we uses oneDNN ReLU to make projection of the perf, it will come to
+~1.8s." Some other data scientist in discussion mentioned that he encouter PReLU
+issue when he tried to optimize the inference pipeline of a popular face-related
+application - PReLU tooks 47 % of time in his application.
+
+Currently in oneDNN we support ReLU activation function in eltwise operation:
+https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html. LeakyReLU can be
+achieved by specifying non zero alpha parameter. PReLU is specific variant of
+LeakyReLU, where alpha is learnable parameter. It cames with two flavours:
+- channel-wise PReLU - where number of trained alpha parameters is equal to
+channel number
+- channel-shared PReLU - where there is only one trainable alpha
+
+All popular framworks supports PReLU:
+- Caffe - https://caffe.berkeleyvision.org/tutorial/layers/prelu.html
+- Keras/tensorflow - https://www.tensorflow.org/api_docs/python/tf/keras/layers/PReLU
+- PyTorch - https://pytorch.org/docs/master/generated/torch.nn.PReLU.html
+- MxNet - https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.nn.PReLU.html
+
+Example of calculations forward/backward can be found: https://bit.ly/2TXgvzT
+
+## Proposal
+
+1. Option A - create new dedicated primitive PReLU:
+
+``` cpp
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_prelu.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training,
+    /// #dnnl_forward_inference, #dnnl_backward
+    dnnl_prop_kind_t prop_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t data_desc;
+    /// Learnable parameter memory descriptor.
+    dnnl_memory_desc_t weights_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Learnable parameter gradient memory descriptor.
+    dnnl_memory_desc_t diff_weights_desc;
+
+} dnnl_prelu_desc_t;
+```
+For passing learnable parameters reuse existing defines:
+``` cpp
+#define DNNL_ARG_WEIGHTS number
+#define DNNL_ARG_DIFF_WEIGHTS number
+```
+Examples:
+n = 256, h = 7, w = 7 c=128
+m_data = {{256, 7, 7, 128}, f32, nhwc};
+
+Algorithm:
+- whole-tensor (tensorflow wariant) - weights_data = {{256, 7, 7, 128}, f32, nhwc};
+
+### Broadcast semantic
+
+Specifying format tag "any" for the weights tensor will enable broadcast feature.
```

Comment:
Why weights format is a condition for enabling broadcast?

Reply:
Maybe I misunderstood Everist comment above (https://github.com/oneapi-src/oneDNN/pull/737#discussion_r434018000) about specifying behaviour for `format_tag::any`. Generally broadcast feature can be enabled when it will match proposed "broadcastable" definition . If we go that way `format_tag::any` will only let us now that format of weights is exactly the same as source input. @emfomenk Maybe you meant something else about `format_tag::any` that I didn't catch ?

Reply:
I agree with @vpirogov that the broadcast semantic is orthogonal to the format tag (this statement is applicable for the entire library, and not only prelu primitive).

What I asked to clarify was what we would recommended users to use when creating a prelu primitive. Roughly, I would expect the following paragraph:
> The library will recommend user to allow prelu primitive to choose the appropriate weights memory format by passing `weights_md` with `format_tag::any`. This might be important for the full-tensor case, where  ideally the weights should repeat the memory format of the data tensor. However, given that the most typical case will be per-channel (C elements) or per-tensor (1 element) weights, the primitive should be optimized for the plain format that user has. Examples:
> - `prelu_forward::desc(src_md, weights_md={{n, c, h, w}, tag::any}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{1, c, 1, 1}, tag::any}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{1, c, 1, 1}, tag::abcd}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{1, 1, 1, 1}, tag::abcd}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{n, c, h, w}, tag::abcd}})` -- not recommended, since primitive might want to use `tab::aBcd16b`

As I said above, the broadcast semantic is orthogonal to the format. This is purely logical think and related to the logical dimensions only.

Reply:
Ok, thank you for clarifying/ I added paragraph above. You used expression "appropriate weights". Do you see any case where primitive with weights description tagged as any should choose different data format than src_md format ? If no maybe I should specifying explicit that "any will let us now that format of weights is exactly the same as src md". I thought that maybe knowing format of weights can help somehow user with data initialisation or maybe in all cases user initializes the data with same || random values and it's format doesn't matter.

Reply:
> Do you see any case where primitive with weights description tagged as any should choose different data format than src_md format?

I don't, though the data types could be different.

> If no maybe I should specifying explicit that "any will let us now that format of weights is exactly the same as src md".

Let it be an implementation detail that we may want to document later in the "implementation detail" section of the documentation. No need to mention that in RFC though.

Reply:
Resolving the thread. Thanks!